`webpushd` app launch improvements

beidson opened the pull request:
#### 5f64555f4058965c0e6713df2d174e0e12217b97
<pre>
`webpushd` app launch improvements
<a href="https://rdar.apple.com/131365223">rdar://131365223</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=278443">https://bugs.webkit.org/show_bug.cgi?id=278443</a>

Reviewed by Ben Nham.

This allows for targeting app launch in a much more flexible way,
allowing for dictionary based communication with the launched app instead of URL based.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
* Source/WebKit/Platform/spi/ios/FrontBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::pushActionVersionKey):
(WebKit::WebPushD::currentPushActionVersion):
(WebKit::WebPushD::pushActionPartitionKey):
(WebKit::WebPushD::pushActionTypeKey):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _setWebPushActionHandler:]):
(+[_WKWebsiteDataStoreBSActionHandler shared]):
(-[_WKWebsiteDataStoreBSActionHandler setWebPushActionHandler:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h: Copied from Source/WebKit/Platform/spi/ios/BaseBoardSPI.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.mm: Copied from Source/WebKit/Platform/spi/ios/FrontBoardServicesSPI.h.
(+[_WKWebPushAction webPushActionWithDictionary:]):
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
* Tools/MobileMiniBrowser/MobileMiniBrowser/MobileMiniBrowser.entitlements:
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/AppDelegate.m:
(-[AppDelegate application:didFinishLaunchingWithOptions:]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController createWebView]):
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos18.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphoneos18.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator18.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator18.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.

Canonical link: <a href="https://commits.webkit.org/282602@main">https://commits.webkit.org/282602@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4cff810392ff95f33f96e185adac3c48d814e70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63635 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67656 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14243 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65755 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50679 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14523 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9846 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66704 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39890 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55101 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31911 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/63146 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36572 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12480 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13116 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/58076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69352 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7582 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/12370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7615 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55194 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14082 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6314 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41003 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/39634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -29,6 +29,7 @@
 
 #if USE(APPLE_INTERNAL_SDK)
 
+#import <UIKit/UIKit.h>
```
This needs to be guarded by `#if PLATFORM(IOS_FAMILY)`. EWS won't catch this since `ENABLE(WRITING_TOOLS)` isn't true on iOS 17.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
```diff
@@ -27,6 +27,8 @@
 
 #if HAVE(CORE_TELEPHONY)
 
+#include <wtf/URL.h>
```
Can we forward declare?

pxlcoder commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -416,21 +416,52 @@ static void tryCloseRequestConnection(xpc_object_t request)
+            FBSOpenApplicationOptionKeyPayloadOptions : @{ UIApplicationLaunchOptionsSourceApplicationKey : @"com.apple.WebKit.webpushd" },
+        };
+
+        _LSOpenConfiguration *configuration = [[_LSOpenConfiguration alloc] init];
```
`adoptNS`? This looks like a leak.

pxlcoder commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -416,21 +416,52 @@ static void tryCloseRequestConnection(xpc_object_t request)
+    FBSOpenApplicationOptions *options = [FBSOpenApplicationOptions optionsWithDictionary:@{
+        FBSOpenApplicationOptionKeyActivateForEvent: @{ FBSActivateForEventOptionTypeBackgroundContentFetching: @{ } },
+        FBSOpenApplicationOptionKeyActivateSuspended : @YES,
+        FBSOpenApplicationOptionKeyActions : @[action.get()],
```
Nit:
```suggestion
        FBSOpenApplicationOptionKeyActions : @[ action.get() ],
```

bnham commented on Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
```diff
@@ -52,6 +52,8 @@ constexpr auto PlatformWritingToolsResultTable = (PlatformWritingToolsResultOpti
 
 #else
 
+#import <UIKit/UIKit.h>
```
Unified build fix?

beidson replied:
Nope, unified source issue.

bnham commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -252,7 +252,11 @@ WK_UIKITSERVICES_LDFLAGS = $(WK_UIKITSERVICES_LDFLAGS_$(WK_PLATFORM_NAME));
 WK_UIKITSERVICES_LDFLAGS_iphonesimulator = -framework UIKitServices;
 
-FRAMEWORK_AND_LIBRARY_LDFLAGS = -lobjc -framework CFNetwork -framework CoreAudio -framework CoreFoundation -framework CoreGraphics -framework CoreText -framework Foundation -framework ImageIO -framework IOKit -framework IOSurface -framework WebKitLegacy -lnetwork -framework Metal $(WK_APPSERVERSUPPORT_LDFLAGS) $(WK_ACCESSIBILITY_LDFLAGS) $(WK_APPKIT_LDFLAGS) $(WK_APPLE_PUSH_SERVICE_LDFLAGS) $(WK_ ...
+WK_SPRINGBOARDSERVICES_LDFLAGS = $(WK_SPRINGBOARDSERVICES_LDFLAGS_$(WK_PLATFORM_NAME));
```
Will require SpringBoardServices.tbd file to be committed.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -348,6 +350,15 @@ void didExceedMemoryFootprintThreshold(size_t footprint, const String& domain, u
+
+@interface _WKWebsiteDataStoreBSActionHandler : NSObject <_UIApplicationBSActionHandler>
++ (_WKWebsiteDataStoreBSActionHandler *)shared;
+- (void)setWebPushActionHandler:(WKWebsiteDataStore *(^)(_WKWebPushAction *action))handler;
```
I wonder if requiring the WKWebsiteDataStore to be provided synchronously in the return value of the callback might be an issue for some clients, but I think we can worry about that issue later if it ever crops up.

bnham commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -416,21 +416,52 @@ static void tryCloseRequestConnection(xpc_object_t request)
+
+    NSDictionary *settingsInfo = @{
+        pushActionVersionKey(): currentPushActionVersion(),
+        pushActionPartitionKey(): (NSString *)subscriptionSetIdentifier.pushPartition,
```
maybe add an early bailout and `RELEASE_LOG_ERROR` is `pushPartition` is somehow nil

bnham commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -416,21 +416,52 @@ static void tryCloseRequestConnection(xpc_object_t request)
-    configuration.frontBoardOptions = options;
-    configuration.allowURLOverrides = NO;
+    RetainPtr bsResponder = [BSActionResponder responderWithHandler:^void (BSActionResponse *response) {
+        // FIXME: Is there something useful we can do if the response is an error?
```
I think we want to at least `RELEASE_LOG_ERROR` if there is an error.

bnham reviewed (commented):
General patch looks fine to me. Are you going to change the BS action handler to fetch/process pending push messages as necessary? (It should also probably wrap all that fetching/processing work with `beginBackgroundTask:` etc.)

bnham commented on Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
```diff
@@ -33,4 +33,30 @@
 - (void)invalidate;
 @end
 
+@interface BSSettings : NSObject
```
In the `#if USE(APPLE_INTERNAL_SDK)` section, it seems like we might be missing some imports?

beidson replied:
Yes. Whatever they are, they arrive in the compilation units indirectly through some other imports.

bnham reviewed (approved):
nit: changelog list slightly out of date

beidson committed:
`webpushd` app launch improvements
rdar://131365223
https://bugs.webkit.org/show_bug.cgi?id=278443

Reviewed by Ben Nham.

This allows for targeting app launch in a much more flexible way,
allowing for dictionary based communication with the launched app instead of URL based.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Platform/spi/ios/BaseBoardSPI.h:
* Source/WebKit/Platform/spi/ios/FrontBoardServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::pushActionVersionKey):
(WebKit::WebPushD::currentPushActionVersion):
(WebKit::WebPushD::pushActionPartitionKey):
(WebKit::WebPushD::pushActionTypeKey):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _setWebPushActionHandler:]):
(+[_WKWebsiteDataStoreBSActionHandler shared]):
(-[_WKWebsiteDataStoreBSActionHandler setWebPushActionHandler:]):
(-[_WKWebsiteDataStoreBSActionHandler _respondToApplicationActions:fromTransitionContext:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.h: Copied from Source/WebKit/Platform/spi/ios/BaseBoardSPI.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushAction.mm: Copied from Source/WebKit/Platform/spi/ios/FrontBoardServicesSPI.h.
(+[_WKWebPushAction webPushActionWithDictionary:]):
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
* Source/WebKit/UIProcess/Cocoa/PlatformWritingToolsUtilities.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::notifyClientPushMessageIsAvailable):
* Tools/MobileMiniBrowser/MobileMiniBrowser/MobileMiniBrowser.entitlements:
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/AppDelegate.m:
(-[AppDelegate application:didFinishLaunchingWithOptions:]):
* Tools/MobileMiniBrowser/MobileMiniBrowserFramework/WebViewController.m:
(-[WebViewController createWebView]):
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphoneos17.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphoneos18.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphoneos18.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator17.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.
* WebKitLibraries/SDKs/iphonesimulator18.0-additions.sdk/System/Library/PrivateFrameworks/FrontBoardServices.framework/FrontBoardServices.tbd:
* WebKitLibraries/SDKs/iphonesimulator18.0-additions.sdk/System/Library/PrivateFrameworks/SpringBoardServices.framework/SpringBoardServices.tbd: Added.

Canonical link: https://commits.webkit.org/282602@main
